library/std/src/sys/windows/alloc.rs

```diff
@@ -1,61 +1,95 @@
+#![deny(unsafe_op_in_unsafe_fn)]
+
 use crate::alloc::{GlobalAlloc, Layout, System};
 use crate::sys::c;
 use crate::sys_common::alloc::{realloc_fallback, MIN_ALIGN};
 
 #[repr(C)]
 struct Header(*mut u8);
 
+/// # Safety
+///
+/// There must be a `Header` at `ptr.offset(-1)`.
 unsafe fn get_header<'a>(ptr: *mut u8) -> &'a mut Header {
-    &mut *(ptr as *mut Header).offset(-1)
+    // SAFETY: the safety contract must be upheld by the caller
+    unsafe { &mut *(ptr as *mut Header).offset(-1) }
 }
 
+/// # Safety
+///
+/// `ptr`, once aligned, must have space for a Header at `ptr.offset(-1)`.
```

Comment:
This is not a sufficient condition. `ptr` must not only have space, but also be valid for writing. Consider borrowing from https://doc.rust-lang.org/nightly/std/ptr/fn.write.html#safety, for example.

It looks like regardless this function is wrong as it gets `&mut` access to uninitialized memory -- it would be worth fixing that before trying to document unsafety here, I think, likely in a separate PR. Or at least leaving a FIXME comment.

Reply:
Should I just inline the function and avoid the &mut?

Reply:
Change `get_header` to return a `*mut Header` and change `align_ptr` to use `ptr::write` to write the header.